src/jit/lsra.cpp

```diff
@@ -3179,9 +3522,38 @@ LinearScan::buildRefPositionsForNode(GenTree *tree,
     int internalCount = buildInternalRegisterDefsForNode(tree, currentLoc, internalRefs);
 
     // pop all ref'd tree temps
-    for (int useIndex=0; useIndex < consume; useIndex++)
+    GenTreeOperandIterator iterator = tree->OperandsBegin(true);
+    LocationInfoList operandDefs;
+    LocationInfoListNode* operandDefsIterator = operandDefs.End();
+    for (int useIndex = 0; useIndex < consume; useIndex++)
     {
-        LocationInfo locInfo = stack->Pop();
+        if (operandDefsIterator == operandDefs.End())
+        {
+            GenTree* operand;
+            do
+            {
+                assert(iterator != tree->OperandsEnd());
+                operand = *iterator;
+
+                ++iterator;
+            } while (!operand->gtLsraInfo.contributesSources);
```

Comment:
>contributesSources [](start = 42, length = 18)

This code could use some comments. In particular:
- It would be good to elucidate here that the first piece of code is getting the register info for (possibly multiple) registers that are defined by a single operand.
- It is not immediately obvious that, once I have captured `operandDefs`, that I can then free those nodes from the map.
- There is often confusion between uses and defs. It would be good to explicitly remind the reader that these `operandDefs` become the uses of the current node, and that if a node `contributesSources` those are actually its defs. It's a bit confusing, so you might redefine `contributesSources` as something like `definesRegs` - since that's more consistent with the `operandDefs` naming.